src/main/sensors/battery.c

```diff
@@ -355,40 +358,67 @@ void powerMeterUpdate(timeUs_t timeDelta)
     mWhDrawn = mWhDrawnRaw / (3600 * 100);
 }
 
-void sagCompensatedVBatUpdate(timeUs_t currentTime)
+void sagCompensatedVBatUpdate(timeUs_t currentTime, timeUs_t timeDelta)
 {
     static timeUs_t recordTimestamp = 0;
     static int32_t amperageRecord;
     static uint16_t vbatRecord;
+    static uint8_t impedanceSampleCount = 0;
+    static pt1Filter_t impedanceFilterState = PT1_FILTER_INITIALIZER;
+    static pt1Filter_t sagCompVBatFilterState;
+    static batteryState_e last_battery_state = BATTERY_NOT_PRESENT;
+
+    if ((batteryState != BATTERY_NOT_PRESENT) && (last_battery_state == BATTERY_NOT_PRESENT)) {
+        pt1FilterReset(&sagCompVBatFilterState, vbat);
+        pt1FilterReset(&impedanceFilterState, 0);
+    }
 
     if (batteryState == BATTERY_NOT_PRESENT) {
 
         recordTimestamp = 0;
+        impedanceSampleCount = 0;
         powerSupplyImpedance = 0;
         sagCompensatedVBat = vbat;
 
     } else {
 
-        if (cmpTimeUs(currentTime, recordTimestamp) > 20000000)
+        if (cmpTimeUs(currentTime, recordTimestamp) > 500000)
             recordTimestamp = 0;
 
         if (!recordTimestamp) {
             amperageRecord = amperage;
             vbatRecord = vbat;
             recordTimestamp = currentTime;
-        } else if ((amperage - amperageRecord >= 400) && ((int16_t)vbatRecord - vbat >= 10)) {
-            powerSupplyImpedance = (int32_t)(vbatRecord - vbat) * 1000 / (amperage - amperageRecord);
-            amperageRecord = amperage;
-            vbatRecord = vbat;
-            recordTimestamp = currentTime;
-        }
+        } else if ((amperage - amperageRecord >= 200) && ((int16_t)vbatRecord - vbat >= 4)) {
 
-        sagCompensatedVBat = MIN(batteryFullVoltage, vbat + powerSupplyImpedance * amperage / 1000);
+            uint16_t impedanceSample = (int32_t)(vbatRecord - vbat) * 1000 / (amperage - amperageRecord);
+
+            if (impedanceSampleCount <= IMPEDANCE_STABLE_SAMPLE_COUNT_THRESH) {
+                impedanceSampleCount += 1;
+            }
 
+            if (impedanceFilterState.state) {
+                impedanceFilterState.RC = impedanceSampleCount > IMPEDANCE_STABLE_SAMPLE_COUNT_THRESH ? 1.2 : 0.5;
+                pt1FilterApply3(&impedanceFilterState, impedanceSample, timeDelta * 1e-6f);
+            } else {
+                pt1FilterReset(&impedanceFilterState, impedanceSample);
+            }
+
+            if (impedanceSampleCount > IMPEDANCE_STABLE_SAMPLE_COUNT_THRESH) {
+                powerSupplyImpedance = lrintf(impedanceFilterState.state);
```

Comment:
This should be made more generic. Accessing filter internal structure is unreasonable and inherently unsafe. Please consider adding a `pr1FilterGetLastOutput` API or similar.